### PR TITLE
test: migrate build_plan and build_script to snapbox

### PR DESCRIPTION
**Pull request by eth3lbert**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/build_plan.rs`
- `tests/testsuite/build_script.rs`

### CODE REVIEW

File: tests/testsuite/build_plan.rs [hunk 1]
```diff
@@ -142,58 +154,75 @@ fn cargo_build_plan_build_script() {
...
+      ],
+      "package_name": "foo",
+      "package_version": "0.5.0",
+      "program": "rustc",
+      "target_kind": [
+        "custom-build"
+      ]
+    },
+    {
+      "args": [],
```
Comment (epage) [inline tests/testsuite/build_plan.rs:184]: Should we glob this like it did before?

Comment (eth3lbert) [reply tests/testsuite/build_plan.rs:184]: Great catch! I'll fix it.

File: tests/testsuite/build_script.rs [hunk 2]
```diff
@@ -3741,6 +3780,7 @@ fn custom_target_dir() {
     p.cargo("build -v").run();
 }
 
+#[allow(deprecated)]
```
Comment (eth3lbert) [inline tests/testsuite/build_script.rs:3783]: `with_stderr_does_not_contain` is being used in this function.

File: tests/testsuite/build_script.rs [hunk 3]
```diff
@@ -4921,55 +5021,54 @@ fn rerun_if_directory() {
...
 
     let fresh = || {
-        p.cargo("check").with_stderr("[FINISHED] [..]").run();
+        p.cargo("check")
+            .with_stderr_data(
+                "\
+[FINISHED] [..]
+",
+            )
+            .run();
```
Comment (eth3lbert) [inline tests/testsuite/build_script.rs:5035]: I think this is less important compared to the `dirty`, which I'm currently just leaving as-is.  Let me know if we want to rewrite it with expected argument.

**Pull request merged**